sw/device/tests/aon_timer_wakeup_irq_test.c

```diff
@@ -0,0 +1,177 @@
+ *
+ * The `ibex_mcycle_read` function uses the `mcycle` register to count
+ * instructions cycles executed and measure time elapsed between two instants,
+ * however it stops counting when the `wfi` is called. As this test is based on
+ * the `wfi` instruction the best approach then to measure the time elapsed is
+ * to use the mtime register, which is basically attached to rv_timer in the
+ * opentitan. https://docs.opentitan.org/hw/ip/rv_timer/doc/
+ *
+ */
+static void tick_init(void) {
+  CHECK_DIF_OK(dif_rv_timer_init(
+      mmio_region_from_addr(TOP_EARLGREY_RV_TIMER_BASE_ADDR), &rv_timer));
+
+  CHECK_DIF_OK(dif_rv_timer_reset(&rv_timer));
+
+  // Compute and set tick parameters (i.e., step, prescale, etc.).
+  dif_rv_timer_tick_params_t tick_params;
+  CHECK_DIF_OK(dif_rv_timer_approximate_tick_params(kClockFreqPeripheralHz,
+                                                    kTickFreqHz, &tick_params));
+
+  CHECK_DIF_OK(
+      dif_rv_timer_set_tick_params(&rv_timer, kPlicTarget, tick_params));
+
+  CHECK_DIF_OK(dif_rv_timer_counter_set_enabled(&rv_timer, kPlicTarget,
+                                                kDifToggleEnabled));
+}
+
+/**
+ * Read the current rv timer count/tick
+ * @return The current rv timer count.
+ */
+static uint64_t tick_count_get(void) {
+  uint64_t tick = 0;
+  dif_rv_timer_counter_read(&rv_timer, kPlicTarget, &tick);
+  return tick;
+}
+
+/**
+ * Execute the wake up interrupt test
+ */
+static void execute_test(dif_aon_timer_t *aon_timer, uint32_t sleep_time_us) {
+  peripheral = kTopEarlgreyPlicPeripheralUnknown;
+  irq = kDifAonTimerIrqWdogTimerBark;
+  // The wake up time should be `sleepTime_us ±15%`
+  uint32_t sleep_range_h = (float)sleep_time_us * 1.15;
+  uint32_t sleep_range_l = (float)sleep_time_us * 0.85;
+
+  // Setup the timer and wait for the IRQ
+  uint32_t sleep_cycles = (sleep_time_us * kClockFreqAonHz / 1000000);
+  LOG_INFO("Setting wakeup interrupt for %u us (%u cycles)", sleep_time_us,
+           sleep_cycles);
+  // Capture the current tick to measure the time the IRQ will take
+  time_elapsed = tick_count_get();
+  aon_timer_testutils_wakeup_config(aon_timer, sleep_cycles);
+  do {
+    wait_for_interrupt();
+  } while (peripheral != kTopEarlgreyPlicPeripheralAonTimerAon &&
+           time_elapsed < sleep_range_h);
+
+  CHECK(time_elapsed < sleep_range_h && time_elapsed > sleep_range_l,
+        "Timer took %u usec which is not in the range %u usec and %u usec",
+        (uint32_t)time_elapsed, sleep_range_l, sleep_range_h);
+
+  CHECK(peripheral == kTopEarlgreyPlicPeripheralAonTimerAon,
+        "Interrupt from incorrect peripheral: exp = %d, obs = %d",
+        kTopEarlgreyPlicPeripheralAonTimerAon, peripheral);
+
+  CHECK(irq == kDifAonTimerIrqWkupTimerExpired,
+        "Interrupt type incorrect: exp = %d, obs = %d",
+        kDifAonTimerIrqWkupTimerExpired, irq);
+}
+
+/**
+ * External interrupt handler
+ */
+void ottf_external_isr(void) {
```

Comment:
Unconditionally assumes that the correct interrupt came in. Perhaps some of the checks from the `execute_test` are better placed here?

Reply:
I added one check to make sure the correct interrupt fired before acknowledging it.